src/upgradeConfig.js

```diff
@@ -1,110 +1,81 @@
+  if (newName !== undefined) {
+    return formatConfigItem(Object.assign(newConfigItem, { name: newName }));
   }
+
+  return configItem;
+};
+
+function buildConfigCollectionTransformer(configItemType) {
+  const itemTransformer = buildConfigItemTransformer(configItemType);
+  return function(configItems) {
+    configItems = parseConfigCollection(configItems);
+
+    return configItems && configItems.map(itemTransformer).filter(Boolean);
+  };
 }
 
-function changePlugins(config) {
-  let plugins = config.plugins;
+const transformPresets = buildConfigCollectionTransformer('preset');
+const transformPlugins = buildConfigCollectionTransformer('plugin');
+
+function appendFlowIfNeeded(presets) {
+  return presets.find(preset => preset === '@babel/preset-flow' || preset[0] === '@babel/preset-flow')
+    ? presets
+    : presets.concat(['@babel/preset-flow']);
+}
 
-  if (!Array.isArray(plugins) && typeof plugins === 'string') {
-    plugins = config.plugins = config.plugins.split(',').map((plugin) => plugin.trim());
+function upgradeConfigForEnv(config, options) {
+  if (config.presets) {
+    config.presets = options.hasFlow
+      ? appendFlowIfNeeded(transformPresets(config.presets))
+      : transformPresets(config.presets);
   }
 
-  // check if plugins are there
-  if (plugins) {
-    // assume it's an array
-    for (let i = 0; i < plugins.length; i++) {
-      let plugin = plugins[i];
-      const pluginsToReplace = Object.keys(oldPlugins);
-
-      // check if it's a plugin with options (an array)
-      if (Array.isArray(plugin)) {
-
-        if (plugin[0].indexOf('babel-plugin') !== 0 && plugin[0].indexOf('@babel/') !== 0) {
-          plugin[0] = `babel-plugin-${plugin[0]}`;
-        }
-
-        if (pluginsToReplace.includes(plugin[0])) {
-          if (oldPlugins[plugin[0]]) {
-            plugin[0] = oldPlugins[plugin[0]];
-          } else {
-            plugins.splice(i, 1);
-            i--;
-          }
-        }
-      } else {
-        if (plugin.indexOf('babel-plugin') !== 0 && plugin[0].indexOf('@babel/') !== 0) {
-          plugin = `babel-plugin-${plugin}`;
-        }
-        if (pluginsToReplace.includes(plugin)) {
-          if (oldPlugins[plugin]) {
-            plugins[i] = oldPlugins[plugin];
-          } else {
-            plugins.splice(i, 1);
-            i--;
-          }
-        }
-      }
-    }
+  if (config.plugins) {
+    config.plugins = transformPlugins(config.plugins);
   }
 }
 
-module.exports = function upgradeConfig(config, options) {
+module.exports = function upgradeConfig(config, options = {}) {
   config = Object.assign({}, config);
 
-  changePresets(config, options);
-  changePlugins(config);
+  upgradeConfigForEnv(config, options);
 
   if (config.env) {
-    Object.keys(config.env).forEach((env) => {
-      changePresets(config.env[env]);
-      changePlugins(config.env[env]);
+    Object.keys(config.env).forEach(env => {
```

Comment:
`Object.values` is what you'd want to use here.

Reply:
heh, great point. that totally didn't occur to me while moving this code around 😆